### DISCUSSION TITLE
Why "No module named 'torch' "  when running exe packaged by pyinstalller

### DISCUSSION THREAD

**Discussion by Moggot** [pyinstaller/pyinstaller#8464] [Help]
This error does not occur when running the script,but packaged by pyinstaller. i had make a hook for pyinstaller and others ways, but error still occur,

**Comment by rokm** [comment]
This usually happens if you have mixed up python (virtual) environments; i.e., you are running PyInstaller that is installed in a different (virtual) environment than the one you are using for your project (and which, in this case, has `torch` installed).

**Reply by Moggot** [reply]
I used a virtual environment to package the software by conda , and I'm sure there's torch in this environment,

**Reply by rokm** [reply]
What is the output of `where pyinstaller`, then?

**Reply by rokm** [reply]
And what is the PyInstaller command you are using?

**Reply by Moggot** [reply]
output is too long but related to this is the following:
```
15136 INFO: Loading module hook 'hook-torch.py' from 'C:\\Users\\27288\\anaconda3\\envs\\openmmlab\\lib\\site-packages\\_pyinstaller_hooks_contr
ib\\hooks\\stdhooks'...
1127 WARNING: Failed to collect submodules for 'torch.testing._internal.distributed._shard.sharded_tensor' because importing 'torch.testing._int
ernal.distributed._shard.sharded_tensor' raised: ModuleNotFoundError: No module named 'expecttest'
1487 WARNING: NOTE: Redirects are currently not supported in Windows or MacOs.
19567 INFO: hook-torch: this torch build does not depend on MKL...
1082 WARNING: Failed to collect submodules for 'torch.testing._internal.distributed._shard.sharded_tensor' because importing 'torch.testing._int
ernal.distributed._shard.sharded_tensor' raised: ModuleNotFoundError: No module named 'expecttest'
1412 WARNING: NOTE: Redirects are currently not supported in Windows or MacOs.
33050 INFO: Loading module hook 'hook-torchvision.py' from 'C:\\Users\\27288\\anaconda3\\envs\\openmmlab\\lib\\site-packages\\_pyinstaller_hooks
_contrib\\hooks\\stdhooks'...
33074 INFO: Loading module hook 'hook-torchvision.io.image.py' from 'C:\\Users\\27288\\anaconda3\\envs\\openmmlab\\lib\\site-packages\\_pyinstal
ler_hooks_contrib\\hooks\\stdhooks'...
35688 INFO: Loading module hook 'hook-torchaudio.py' from 'C:\\Users\\27288\\anaconda3\\envs\\openmmlab\\lib\\site-packages\\_pyinstaller_hooks_
contrib\\hooks\\stdhooks'...
C:\Users\27288\anaconda3\envs\openmmlab\lib\site-packages\torchaudio\backend\utils.py:62: UserWarning: No audio backend is available.
  warnings.warn("No audio backend is available.")
C:\Users\27288\anaconda3\envs\openmmlab\lib\site-packages\torchaudio\backend\utils.py:62: UserWarning: No audio backend is available.
  warnings.warn("No audio backend is available.")
C:\Users\27288\anaconda3\envs\openmmlab\lib\site-packages\caffe2\__init__.py:5: UserWarning: Caffe2 support is not fully enabled in this PyTorch
 build. Please enable Caffe2 by building PyTorch from source with `BUILD_CAFFE2=1` flag.
  warnings.warn("Caffe2 support is not fully enabled in this PyTorch build. "
C:\Users\27288\anaconda3\envs\openmmlab\lib\site-packages\caffe2\proto\__init__.py:17: UserWarning: Caffe2 support is not enabled in this PyTorc
h build. Please enable Caffe2 by building PyTorch from source with `BUILD_CAFFE2=1` flag.
  warnings.warn('Caffe2 support is not enabled in this PyTorch build. '
C:\Users\27288\anaconda3\envs\openmmlab\lib\site-packages\caffe2\proto\__init__.py:17: UserWarning: Caffe2 support is not enabled in this PyTorc
h build. Please enable Caffe2 by building PyTorch from source with `BUILD_CAFFE2=1` flag.
  warnings.warn('Caffe2 support is not enabled in this PyTorch build. '
C:\Users\27288\anaconda3\envs\openmmlab\lib\site-packages\caffe2\python\__init__.py:9: UserWarning: Caffe2 support is not enabled in this PyTorc
h build. Please enable Caffe2 by building PyTorch from source with `BUILD_CAFFE2=1` flag.
  warnings.warn('Caffe2 support is not enabled in this PyTorch build. '
C:\Users\27288\anaconda3\envs\openmmlab\lib\site-packages\caffe2\proto\__init__.py:17: UserWarning: Caffe2 support is not enabled in this PyTorc
h build. Please enable Caffe2 by building PyTorch from source with `BUILD_CAFFE2=1` flag.
  warnings.warn('Caffe2 support is not enabled in this PyTorch build. '
C:\Users\27288\anaconda3\envs\openmmlab\lib\site-packages\caffe2\python\__init__.py:9: UserWarning: Caffe2 support is not enabled in this PyTorc
h build. Please enable Caffe2 by building PyTorch from source with `BUILD_CAFFE2=1` flag.
  warnings.warn('Caffe2 support is not enabled in this PyTorch build. '
C:\Users\27288\anaconda3\envs\openmmlab\lib\site-packages\torchaudio\backend\utils.py:62: UserWarning: No audio backend is available.
  warnings.warn("No audio backend is available.")
14941 INFO: Loading module hook 'hook-torch.py' from 'C:\\Users\\27288\\anaconda3\\envs\\openmmlab\\lib\\site-packages\\_pyinstaller_hooks_contr
ib\\hooks\\stdhooks'...
1133 WARNING: Failed to collect submodules for 'torch.testing._internal.distributed._shard.sharded_tensor' because importing 'torch.testing._int
ernal.distributed._shard.sharded_tensor' raised: ModuleNotFoundError: No module named 'expecttest'
1470 WARNING: NOTE: Redirects are currently not supported in Windows or MacOs.
19369 INFO: hook-torch: this torch build does not depend on MKL...
1114 WARNING: Failed to collect submodules for 'torch.testing._internal.distributed._shard.sharded_tensor' because importing 'torch.testing._int
ernal.distributed._shard.sharded_tensor' raised: ModuleNotFoundError: No module named 'expecttest'   etc..
```

**Reply by Moggot** [reply]
The exported mian.spec
 are as follows:
```
a = Analysis(
    ['main.py'],
    pathex=['D:\\PythonProject\\Termor\\Tremor_2024_5_11\\clinic_judgement\\judgement_core',
    'D:\\PythonProject\\Termor\\Tremor_2024_5_11\\clinic_judgement\\algorithm_core',
    'D:\\PythonProject\\Termor\\Tremor_2024_5_11\\clinic_judgement\\yolov7',
    ],
    binaries=[],
    datas=[('./images/','images'),('./font/','fonts'),('./clinic_judgement/yolov7/','clinic_judgement/yolov7/')
    ,('./clinic_judgement/weights/','clinic_judgement/weights'),('./clinic_judgement/mmpose/','mmpose'),('./config/','config'),
    ],
    hiddenimports=['mmcv._ext','pytorch',],
    hookspath=['C:/Users/27288/anaconda3/envs/openmmlab/Lib/site-packages/PyInstaller/hooks/hook-torch.py'],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    noarchive=False,
    optimize=0,
)



pyz = PYZ(a.pure)

exe = EXE(
    pyz,
    a.scripts,
    [],
    exclude_binaries=True,
    name='main',
    debug=False,
    bootloader_ignore_signals=False,
    strip=False,
    upx=True,
    console=False,
    disable_windowed_traceback=False,
    argv_emulation=False,
    target_arch=None,
    codesign_identity=None,
    entitlements_file=None,
)
coll = COLLECT(
    exe,
    a.binaries,
    a.datas,
    strip=False,
    upx=True,
    upx_exclude=[],
    name='main',
)
```

**Reply by Moggot** [reply]
60054 INFO: Analyzing hidden import 'pytorch'
60055 ERROR: Hidden import 'pytorch' not found

I found this output just now

**Reply by rokm** [reply]
> 60054 INFO: Analyzing hidden import 'pytorch' 60055 ERROR: Hidden import 'pytorch' not found
> 
> I found this output just now

This is red herring. You have added `pytorch` to hiddenimports in the spec file, but `pytorch` is not a valid importable name (the importable package is called `torch`), hence the warning/error message. But it has no real effect.

**Reply by Moggot** [reply]
Thank, i'll try

**Reply by rokm** [reply]
Can you post the full error traceback that you get at run-time?

**Reply by rokm** [reply]
Because seeing that the torch hooks are being ran, I find it hard to believe that `torch` module itself would be missing at run-time (one of submodules, perhaps, but not the main one)...

**Reply by Moggot** [reply]
error traceback is following:
INFO:utils.torch_utils:Model Summary: 314 layers, 36481772 parameters, 6194944 gradients
torch\functional.py:478: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at  C:\actions-runner\_work\pytorch\pytorch\builder\windows\pytorch\aten\src\ATen\native\TensorShape.cpp:2895.)
  return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]
Fusing layers... 
RepConv.fuse_repvgg_block
RepConv.fuse_repvgg_block
RepConv.fuse_repvgg_block
IDetect.fuse
 Convert model to Traced-model... 
 traced_script_module saved! 
 model is traced! 

4 for training 
1 for testing
Start transform please wait ...
  0%|          | 0/4 [00:00<?, ?it/s]100%|██████████| 4/4 [00:00<00:00, 4013.69it/s]
  0%|          | 0/1 [00:00<?, ?it/s]100%|██████████| 1/1 [00:00<?, ?it/s]
  0%|          | 0/4 [00:00<?, ?it/s]100%|██████████| 4/4 [00:00<?, ?it/s]
  0%|          | 0/1 [00:00<?, ?it/s]100%|██████████| 1/1 [00:00<00:00, 1003.18it/s]

Traceback (most recent call last):
  File "D:\PythonProject\Termor\Tremor_2024_5_11\dist\main\_internal\clinic_judgement\yolov7\train.py", line 12, in <module>
    from torch.distributed import broadcast, init_process_group, destroy_process_group
ModuleNotFoundError: No module named 'torch'

**Reply by Moggot** [reply]
hiddenimports=['mmcv._ext','torch',]
the bug still exist

**Reply by rokm** [reply]
How is that `D:\PythonProject\Termor\Tremor_2024_5_11\dist\main_internal\clinic_judgement\yolov7\train.py` ran? It does not seem to be part of the frozen executable itself - is this ran in some external python interpreter process?

Because if `torch` was missing from the frozen applicaiton itself, then you wouldn't be getting other torch-related output at all.

**Reply by Moggot** [reply]
Maybe i didnt import the submodule somewhere, i import the whole torch, so the error occur in the “import torch"

**Reply by rokm** [reply]
Yes, but the error does not seem to originate from the "main" application. This seems to be using torch distributed framework, and the traceback suggests that the `train.py` script is ran in external (sub)process. The question is, what python interpreter does that use, because PyInstaller uses embedded interpreter that is not available to arbitrary subprocesses.

**Reply by Moggot** [reply]
Yeah,train.py is the external process of main application. Main thread maybe use the  interpreter for the virtual environment ,and external process use the computer's default interpreter

**Reply by Moggot** [reply]
Is there any way for external programs to use the same interpreter as the main program

**Reply by rokm** [reply]
The main application uses embedded interpreter with collected packages, but if external process uses the computer's default interpreter, then those packages are naturally not available there (unless installed). 

But if you need external interpreter, then there is no point in using PyInstaller in the first place, since your application is not portable (won't work if target machine has no python interpreter, or does not have torch and other dependencies installed).

**Reply by rokm** [reply]
> Is there any way for external programs to use the same interpreter as the main program

No, because embedded/isolated interpreter is used, which is inaccessible "from outside". You *could* route them through the main executable, but you will need to adjust the entry-point script (main.py) to identify subprocess (e.g., via command-line arguments), and then divert the program flow accordingly. Then you would spawn your subprocesses using `sys.executable` with afore-mentioned command-line arguments.

We actually do this for `multiprocessing`, but whether you can do it in your case probably depends on how torch distributed framework works. (Or maybe rather, are you spawning that `train.py` script process yourself, or is torch.distributed doing it for you).

**Reply by Moggot** [reply]
I specified the interpreter for the virtual environment when I ran the external program, then pyinstaller found the module successfully. Thanks a lot, I will try to see if the exported file would work without an interpreter.
